src/hooks/useFitsScreen.ts

```diff
@@ -0,0 +1,183 @@
+      // Whether the bottom half of the target element fits between the vertical
+      // center of the reference element and the bottom edge of the screen.
+      fitsBelow: boolean;
+      // Whether the left half of the target element fits between the horizontal
+      // center of the reference element and the left edge of the screen.
+      fitsLeft: boolean;
+      // Whether the right half of the target element fits between the
+      // horizontal center of the reference element and the right edge of
+      // the screen.
+      fitsRight: boolean;
+    };
+  };
+};
+
+/**
+ * A hook to determine if an element fits on the screen.
+ * @param targetNode The element to try and fit on the screen.
+ * @param referenceNode The element to use to position the target.
+ * @param callback The function to call when updating fitment info.
+ * @param spacer An additional space to leave between the target and reference.
+ * @param shouldCheck Whether the fitment info should be being checked.
+ * @param fromMouse Whether the target should be being positioned in relation
+ *                  to the mouse. In this case refernceNode will be used to
+ *                  listen for mouseover events.
+ */
+export const useFitsScreen = (
+  targetNode: HTMLElement,
+  referenceNode: HTMLElement,
+  callback: (fitsScreen: FitsScreen) => void,
+  spacer = 0,
+  shouldCheck = true,
+  fromMouse = false
+): void => {
+  const htmlRef = useRef<HTMLElement>(document.querySelector("html"));
+
+  const update = useCallback(
+    (evt?) => {
+      let referenceCoordinates: {
+        height: number;
+        left: number;
+        top: number;
+        width: number;
+      };
+      if (fromMouse) {
+        if (evt) {
+          referenceCoordinates = {
+            // The mouse is a single point so use 0 for the height and width.
+            height: 0,
+            left: evt.x || 0,
+            top: evt.y || 0,
+            width: 0,
+          };
+        }
+      } else if (referenceNode) {
+        referenceCoordinates = referenceNode.getBoundingClientRect();
+      }
+      if (shouldCheck && targetNode && referenceCoordinates) {
+        const {
+          height: targetHeight,
+          width: targetWidth,
+        } = targetNode.getBoundingClientRect();
+        const {
+          height: referenceHeight,
+          left: referenceLeft,
+          top: referenceTop,
+          width: referenceWidth,
+        } = referenceCoordinates;
+        const referenceBottom = referenceTop + referenceHeight;
+        const referenceRight = referenceLeft + referenceWidth;
+        const referenceCenterX = referenceLeft + referenceWidth / 2;
+        const referenceCenterY = referenceTop + referenceHeight / 2;
+        const screenLeft = htmlRef.current.scrollLeft;
+        const screenRight = window.innerWidth - htmlRef.current.scrollLeft;
+        const screenTop = 0;
+        const screenBottom = window.innerHeight;
+
+        callback({
```

Comment:
There is a lot of repeated math here, can they be abstracted out into variables to make this easier to read (and compress better)?

Reply:
Done, every repeated calculation is now a variable.